### PR TITLE
[TwigBundle] Fix bug where namespaced paths don't take parent bundles in account

### PR DESCRIPTION
**Pull request by wesleylancel**
| Q | A |
| --- | --- |
| Branch? | 2.7 |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #6919 |
| License | MIT |
| Doc PR |  |

Currently namespaced paths for templates such as `{% extends '@App/Layout/layout.html.twig' %}` do not work with bundles that have overruled templates using the `getParent()` method in another bundle. See attached ticket. This change prepends the path of the bundle implementing `getParent()` to the paths of the namespace of bundle returned as a parent.

### CODE REVIEW

**Comment by jakzal** [comment]
We'll need a test case for this.

**Comment by wesleylancel** [comment]
@jakzal I modified the TwigExtensionTest to include a test case for this situation.

**Comment by wesleylancel** [comment]
AppVeyor ran on an outdated commit, might have to be restarted.

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php
Comment (ro0NL) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: This is not really a mock :) what about `$object`?

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php
Comment (ro0NL) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: I guess `null !== $parent` is more verbose/consistent.

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php
Comment (ro0NL) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: What about `normalizeBundleName`?

File: src/Symfony/Bundle/TwigBundle/Tests/DependencyInjection/Fixtures/Bundle/TwigResourcesBundle/TwigResourcesBundle.php
Comment (ro0NL) [inline src/Symfony/Bundle/TwigBundle/Tests/DependencyInjection/Fixtures/Bundle/TwigResourcesBundle/TwigResourcesBundle.php]: Perhaps `ChildTwigBundle` is more verbose..

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php
Comment (wesleylancel) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: I'll rename the variable. Don't think there are any options to retrieve the parent in this situation.

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php
Comment (ro0NL) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: That makes sense ;-) but do you agree on this approach? Im not sure how much we can rely on the default ordering of `kernel.bundles`, i guess we should correctly order them here for now.

Comment (wesleylancel) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: Paths of child bundles will be prepended so they'll automatically come earlier than the original bundle's. Extra ordering should not be necessary in my opinion?

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php
Comment (ro0NL) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: Maybe you could call `registerBundles` once =/

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php
Comment (ro0NL) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: If the order of `kernel.bundles` is correct (in a way that parents come before childs), then this is approach works. Basically i think we need a test that proves a more complex bundle hierarchy works as expected.

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php
Comment (wesleylancel) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: I agree, and this won't work with a third level (ChildChildTwigBundle). How would you recommend calling `registerBundles` here? Is there any way to reach the kernel?

Comment (ro0NL) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: `kernel` service is out of scope, so i guess not. Maybe you can kick in at `TwigBundle::createContainerExtension`, without breaking BC.

This is somewhat related to #18563 in terms of improving the ecosystem.

Comment (ro0NL) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: Point is i think you need to traverse all possible parents, and register them for the current bundle `$dir`, so each parent works as expected.

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php
Comment (ro0NL) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: The order is irrelevant, point is i think you need to traverse all possible parents, and register them for the current bundle `$dir`, so each parent works as expected.

Comment (wesleylancel) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: Correct, I missed that. That seems impossible with the current code unless I can retrieve the complete bundleMap.

Comment (ro0NL) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: Which you have... `$container->getParameter('kernel.bundles')` is a map of short- to classname.

Comment (wesleylancel) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: The only way I could think of making this work is indeed, recursively instantiating the parent class, getting the parent bundle name, retrieve its class from the list of bundles, and so on. Is that what you would recommend?

Comment (ro0NL) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: Yes, with performance in mind. Each bundle in the map should only be initialized once. Just keep in mind you're in a loop.

Comment (ro0NL) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: Also if you have A>B>C, asking the parents for A should resolve the parents for B also. So if you ask for that later it's there.

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php
Comment (jakzal) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: I'm afraid we cannot instantiate the bundle class like this. Some bundles might have overridden the constructor (like the [JMSDebuggingBundle](https://github.com/schmittjoh/JMSDebuggingBundle/blob/master/JMSDebuggingBundle.php#L33)).

Comment (ro0NL) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: What about this; as bundles are initialized before the container is ([ref](https://github.com/symfony/symfony/blob/973f29551255cac22b87e9e82ff93a280c9959c6/src/Symfony/Component/HttpKernel/Kernel.php#L114)) we can easily inject this kind of information. Currently `kernel.bundles` giving you only the FQCN is not flexible. We sometimes need the path, parent etc. So im thinking a special VO that replaces the FQCN is worth it... for now as additional parameter perhaps, but eventually to deprecate `kernel.bundles` being a list of FQCN's...

I would like this a lot!

Comment (wesleylancel) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: @ro0NL Are you suggesting inject the entire `bundleMap` as a parameter after it has been initialized? This would also remove the need to initialize any classes. /cc @jakzal

Comment (ro0NL) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: See #19594

Comment (wesleylancel) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: Shouldn't the existing `bundleMap` property on the `Kernel` be able to provide the necessary data?

Comment (ro0NL) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: Good point, havent noticed that one. Lets discuss further implementation details in #19594.

Comment (xabbuh) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: What about using reflection and `newInstanceWithoutConstructor()` here?

**Comment by xabbuh** [comment]
@wesleylancel Do you think you can finish this one?

**Comment by wesleylancel** [comment]
@xabbuh: I wish, as this is something that would be extremely useful for us. But, as pointed out by @ro0NL, bundles can extend infinitely A > B > C > D. My PR currently only checks for one level of extending. Ideally, I would be able to access the bundle map that is part of the kernel, but that's not available at this stage yet. If you could point me towards a possible solution I'd be happy to adjust my code accordingly.

**Comment by ro0NL** [comment]
Seeing the related ticket this goes back to 2013. And quite some issues are involved. I've not experienced it myself, but the template problem seems real.

Either way allowing kernel introspection would be _extremely useful_ for everybody. Preferably with an immutable kernel. Practically we can move [this](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpKernel/Kernel.php#L493) a few lines up. Relying on a synthetic service before compilation.

We also can expose bundles thru parameters, we only need stringish values here. `BundleInterface` only defines name, path and namespace anyway. So passing a stringish bundle map could work. Any further details can be provided with `Bundle::build`.

But mayve @ogizanagi  has more idea's? As i also liked his approach with a contextual builder. Maybe it can be simplified.

**Comment by ogizanagi** [comment]
Unfortunately #19646 was my only attempt to provide a naive feature for sharing some contextual information while the container is building, avoiding the need to rely on pseudo tmp synthetic services. Along with the `BootingKernel` and `BootingBundle` classes, it allowed to solves issues related to this one as a demonstration of the possibilities.
But despite the implementation is simple, it somehow looks overkilled for the few needs.

As you suggested, maybe exposing a simple representation of the bundle map and bundles informations in a kernel parameter could be the way. But I don't remember all the ins and outs and if it was easily feasible at this point. :/

**Comment by ro0NL** [comment]
> But despite the implementation is simple, it somehow looks overkilled for the few needs.

Agree:+1: I feel the same about my own attempts now. Eventually having builder with context can be done in userland easily. But it doesnt solve it for SF.

If there arent any other insights, i'd say go for `%kernel.bundle_map%` :+1:

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php
Comment (chalasr) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: @xabbuh some bundles make use of constructor arguments in their `getPath()` method. 
I tried to solve a similar issue in https://github.com/symfony/symfony/pull/18579

Comment (xabbuh) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: That's true, but the only ways to solve that issue completely would be to either pass bundles to extension classes or to write some logic that determines the paths during cache warmup.

Comment (chalasr) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: Couldn't a parameter returning the bundle map be envisaged?

Comment (xabbuh) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: The issue is that you still would not have access to actual instances of the bundles (so this would not solve https://github.com/symfony/symfony/pull/19586#r94117685).

**Comment by chalasr** [comment]
>  i'd say go for %kernel.bundle_map%

Not sure it's doable. Putting `Kernel::$bundleMap` in a new kernel parameter would just lead to

>  Unable to dump a service container if a parameter is an object or a resource.

**Comment by ro0NL** [comment]
It's a stringish map, that's actually cached :) we're not passing bundle instances around here.

```
array(
  'name' => ['parent' => 'name2', 'namespace' => 'NS', 'path' => '/bundles/name' ],
  'name2' => //
)
```

Something like that i guess.

**Comment by ro0NL** [comment]
Ie. looking at the issues, im not really sure we need true bundle instances here. Only metadata from bundle introspection. To build things properly.

**Comment by chalasr** [comment]
See #21113

**Comment by wesleylancel** [comment]
Nice @chalasr! I've subscribed to the PR and will update this one when it's merged.

**Comment by xabbuh** [comment]
@wesleylancel *21113 has been merged.

**Comment by wesleylancel** [comment]
@xabbuh Thanks for the heads up. Looking into rewriting this PR using the new parameter.

**Comment by ogizanagi** [comment]
Is it still needed? @chalasr made the changes in the TwigExtension. What remains to do?

**Comment by wesleylancel** [comment]
@ogizanagi I don't think that fixes registering extra paths for parent bundles.

**Comment by chalasr** [comment]
>  I don't think that fixes registering extra paths for parent bundles.

Indeed, #21113 fixed the case for bundles overriding `getPath()`, handling bundle hierarchy still needs to be fixed

**Comment by wesleylancel** [comment]
@xabbuh @chalasr Just committed a first attempt using the new parameter. To the contrary of the earlier version this should take multiple levels of nested bundle into account. Feedback welcome.

**Comment by chalasr** [comment]
@wesleylancel Looks good to me 👍  Not sure why but it seems you have a conflict

**Comment by fabpot** [comment]
You should probably change the branch target to 2.7 instead of master.

**Comment by wesleylancel** [comment]
Done.

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php
Comment (xabbuh) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: should be `$bundleMetadata['path']`

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php
Comment (xabbuh) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: `$name` should be `$parentName`

**Review by xabbuh — CHANGES_REQUESTED** [review]

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php
Comment (wesleylancel) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: @xabbuh: `$bundleMetadata` contains the entire array of bundles, `$parentBundle` is correct I think?

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php
Comment (wesleylancel) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: @xabbuh The path should be registered with the bundle namespace that extends the parent bundle, so I think this should be allright too.

Comment (xabbuh) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: But you want to register the path of the current bundle as a resource for the parent bundle's namespace.

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php
Comment (xabbuh) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: Ah sorry, I meant `$bundleMetadata[$name]['path']`.

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php
Comment (wesleylancel) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: Ah, yep, it's the wrong way around indeed. Will fix.

**Comment by wesleylancel** [comment]
@xabbuh I think it should be the right way around now :)

**Comment by chalasr** [comment]
@wesleylancel The calls of `realpath()` need to be removed [here](https://github.com/symfony/symfony/pull/19586/files#diff-00bc5ea153482178966f6617538a8e2eR282) and [here](https://github.com/symfony/symfony/pull/19586/files#diff-00bc5ea153482178966f6617538a8e2eR287) for the appveyor build to pass. Once fixed I would say we're good here

**Comment by wesleylancel** [comment]
@chalasr Thanks, fixed.

**Comment by dunglas** [comment]
Looks good to met.

👍

**Comment by xabbuh** [comment]
👍 I checked this with the issues described in FriendsOfSymfony/FOSUserBundle#2376 and FriendsOfSymfony/FOSUserBundle#2378 and I can confirm that this patch fixes them.

Status: Reviewed

**Review by xabbuh — APPROVED** [review]

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php
Comment (stof) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: Does this handle a chain of parents ?

**Comment by kmelia** [comment]
@xabbuh great news! Could you tell us on which branches this patch will be applied ?

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php
Comment (stof) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: Prepending paths looks wrong to me here, as it would overwrite user-configured paths for the same namespace too.
What you should do is collecting the child bundles for each bundle, and then adding paths for child bundles before adding the path of the bundle itself

**Comment by chalasr** [comment]
@kmelia Once merged, this will be available in the next patch releases of each maintained branch (i.e. from 2.7 to 3.2)  and the current development version (3.3).

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php
Comment (wesleylancel) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: @stof: Yes, this handles that. Should be covered with the changed test.

Comment (wesleylancel) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: @stof: Did not now about that. That'll require some change to adding the normal paths too, I'll see how to best fix this. Is there an easy way to cover this in the test?

**Comment by kmelia** [comment]
@chalasr thanks!

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php
Comment (stof) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: Add a path in the user-configured paths for the same namespace, and ensure that it ends up as the first path in the loader config for this namespace

File: src/Symfony/Bundle/TwigBundle/Tests/DependencyInjection/TwigExtensionTest.php
Comment (stof) [inline src/Symfony/Bundle/TwigBundle/Tests/DependencyInjection/TwigExtensionTest.php]: The way these tests are written does not ensure that paths are added in the right order for each namespace. there is nothing guaranteeing that the combinations of appended paths and prepended paths is actually giving us the right priority.

What we should do to have reliable tests is rebuilding the internal structure of ``Twig_Loader_Filesystem::$path`` (i.e. the result of the configuration) and ensuring that it looks good for each namespace.

Comment (wesleylancel) [inline src/Symfony/Bundle/TwigBundle/Tests/DependencyInjection/TwigExtensionTest.php]: @stof: I think whatever gets prepended last will have the most priority. If the array above is in this order, the correct path will end up on top. However, this doesn't keep those user-configured paths in mind. I'm not sure where these get added to the loader, but if they get prepended after the bundle paths get prepended, it should work all right? Not sure if that's testable in the existing test though.

If the Twig internals of that loader do need to be changed, I don't think this can be merged before that is resolved?

Comment (wesleylancel) [inline src/Symfony/Bundle/TwigBundle/Tests/DependencyInjection/TwigExtensionTest.php]: @stof: What if we move these lines https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php#L84-L90 to here https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php#L109 and change the method call to `prependPath` instead of `addPath`? This should make sure that user-defined paths from config take precedence over anything else.

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php
Comment (ro0NL) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: `prependParentPaths` was confusing at first. Maybe keep actual path registration explicit in `load()`

```php
while ($parent) {
   // add parent
}

// add self
```

Basically we dont need any path prepending or recursion, only adding paths with the right calling order.

File: src/Symfony/Bundle/TwigBundle/Tests/DependencyInjection/TwigExtensionTest.php
Comment (stof) [inline src/Symfony/Bundle/TwigBundle/Tests/DependencyInjection/TwigExtensionTest.php]: the issue is that the way the test is written does not allow to quickly validate what the final order will be when looking at it. This is why I think the test should not be written this way. We don't want to assert 2 separate lists, as we care about a single list (the one being used at runtime).

thus, the test is even confusing when looking at it, because the list of ``prependPath`` here is in reverse order compared to the order of paths in the loader (as prepending adds at the beginning while you append in ``$prependPaths``), making it even harder to validate/understand the assertion.

Comment (wesleylancel) [inline src/Symfony/Bundle/TwigBundle/Tests/DependencyInjection/TwigExtensionTest.php]: @stof I'll see if I can collect all the paths before actually adding them to the loader so they can all be tested together in order.

**Comment by wesleylancel** [comment]
@stof I've added a separate commit so it's easier to see what I've changed (can be squashed later). A hierarchy of bundles is now built first, after that the paths are added to the loader. I'm not sure if I could simplify the code, especially the building of the hierarchy.

I could improve the test by asserting the array per namespace if that helps for readability?

**Comment by xabbuh** [comment]
I think an easier implementation could be to have two kids. The first would collect path information for bundles we override without actually registering them. The second would then first register the additional paths from child bundles followed by paths the bundle provides itself.

**Comment by wesleylancel** [comment]
@xabbuh: Isn't that what I'm doing now? First the hierarchy is built and then the paths are registered.

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php
Comment (stof) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: Note to maintainers: when merging to newer branches, we need to take care of the FileExistenceResource objects here (git conflicts should remind us about this btw)

**Comment by stof** [comment]
👍

**Comment by dunglas** [comment]
👍

**Comment by fabpot** [comment]
Thank you @wesleylancel.

**Comment by homersimpsons** [comment]
@fabpot Do you have any hint about the new release (containing this fix) date?

**Comment by chalasr** [comment]
@homersimpsons today :) see https://github.com/symfony/symfony/pull/21258

**Pull request closed without merging**
